### DISCUSSION TITLE
Persistent Islands

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#307] [Proposal]
# Persistent Islands 

## Summary

This is a spin on the "SPA mode" idea we've been kicking around for a long time. The key difference is that instead of turning on SPA mode for the entire site, persistent islands is a way to mark DOM that should persist between page navigations.

This is holistic with our usage based opt-in approach to client JS and allows for sites that are a mixture of MPA and SPA. When the user lands on a page without any persistent islands the next navigation will be a full-page MPA-style navigation. When they land on a page with persistent islands, the next navigation will be SPA-style, with the marked islands persisting.

## Example

To mark an island as being persistent use the `navigation:persist` attribute like so:

```astro
---
import Header from '../components/Header.jsx';
---
<html>
<head>
  <title>E-commerce website</title>
</head>
<body>
  <Header client:idle navigation:persist />
  <main>
    <h1>Products</h1>
    <p>Some main content that does not persist.</p>
  </main>
</body>
</html>
```

Note that there's probably no reason that `navigation:persist` can't be applied to non-interactive content such as Astro components:

```astro
---
import Sidenav from './Sidenav.astro';
---

<Sidenav navigation:persist />
```

## Motivation

Astro users have been asking for a way to persist content on pages for a long time. We have played around with many "SPA mode" approaches.

This is an attempt to bring some of the benefits of SPA mode without giving up the simplicity of multi-page architecture, and to pair it with Astro's opt-in approach to more advanced use cases.

## Design

This likely would be implemented very similarly to previous SPA-mode proposals. It might be nice to, instead of diffing the page, simply stitch together DOM around the islands. Need to experiment to see how possible/practical that would be.

## Side note

Thinking about this idea has me also thinking about the previous [server islands](https://github.com/withastro/rfcs/discussions/116) idea.

Imagine you could mark a component as `server:dynamic`. This would statically generate all of the rest of the page except for those components. In effect you get "per-page SSR vs SSG" using this method, but again using usage-based opt-in rather than page-level or site-level configuration.

**Comment by matthewp** [comment]
A somewhat quick and dirty demo:

https://user-images.githubusercontent.com/361671/190869418-5d64719c-6ae3-4831-b57d-e25a197b058a.mp4

**Comment by tony-sull** [comment]
Notes from the Sept 20 Community Call

- Any concerns with scripts and styles being removed during a page navigation?
- Lots of love in the chat, lots of excitement every time SPA mode and similar features come up on a call
- What are the next steps to move this idea forward?

**Comment by tanepiper** [comment]
This is a really nice proposal that fits very nicely without needs as we use a lot of SSG.  We chose the following architecture:

Use Astro itself sparingly as templates for generated pages, on each page using a template that loads a header, sidebar and footer. The first two should be persistent across all pages.

To solve this we build an Observable store with `sessionStorage` and API backends, so for each page transition we check sessionStorage first and get the state, otherwise, we fetch the data on the first loading of the home page for the header, which contains the user button.

For the sidebar, it's about keeping the navigation state open and showing the user's name from the shared user state at the top.

With future opt-in for ssg vs ssr, we can see us moving to hybrid but again needing this approach across multiple pages.

**Comment by walpolea** [comment]
Love the ideas here! There is an added feature MPA-to-SPA routing libraries like turbo or taxi.js give you with the idea of persistent elements: seamless page transitions. **I'm curious, would this proposal include the event lifecycle it takes to enable seamless transitions?**

I've just finished a project using [taxi.js](https://taxi.js.org/) and really enjoyed their Transitions API which allows you to define multiple transition types and animate DOM elements from the leaving page content to the entering page content.

If you do plan to include this feature in the scope I could see there being an issue with deciding to choose which navigation style based on the initial existence of persistent islands, because you may want to persist an item in real-time (maybe because you clicked it), changing it to being persistent right before you navigate away.

**Comment by MrSunshyne** [comment]
This is something that MPAs lack, and kind of pain to implement in astro. The Shared Element Transition API still sits behind a flag on chrome today and nonexistent on other major browsers. 
Would be really cool to have the opinion on the astro maintainers on this topic.

**Comment by natemoo-re** [comment]
This proposal has been accepted and promoted to **Stage 2** as **Client-side Routing** (https://github.com/withastro/roadmap/issues/532).

Please continue discussion on https://github.com/withastro/roadmap/issues/532.
